solutions/platinum/cses-1706.mdx

```diff
@@ -0,0 +1,131 @@
+---
+id: cses-1706
+source: CSES
+title: School Excursion
+author: Brebenel Mihnea
+---
+
+## Explanation
+
+The solution is based on DSU ([Disjoint Set Union](/gold/dsu)).
+For example: If 2 and 3 want the same thing and 3 and 4 want the same thing it means that 2, 3 and 4 want the same thing. After we determine the components we need to find out every combination of them. To do that we use a knapsack DP (you can find a good tutorial [here](/gold/knapsack)). But it would take $\mathcal{O}(N ^ 2)$ which will cause TLE. In order to speed it up we use a bitset as a knpasack. The final complexity is $\mathcal{O}(N\log^2N)$.
```

Comment:
```suggestion
For example: If 2 and 3 want the same thing and 3 and 4 want the same thing it means that 2, 3 and 4 want the same thing. After we determine the components we need to find out every combination of them. To do that we use a knapsack DP (you can find a good tutorial [here](/gold/knapsack)). But it would take $\mathcal{O}(N ^ 2)$ which will cause TLE. In order to speed it up we use a bitset as a knapsack. The final complexity is $\mathcal{O}(N\log^2N)$.
```

Reply:
also, this time complexity seems wrong

Reply:
still looks wrong, how did you come up with N log N?

Reply:
It's $\mathcal{O}(N) $ amortized complexity. I think that log N comes from shifting bits and the DSU.

Reply:
I don't think so. could you explain your reasoning?

Reply:
Well, in the worst case every kid could be a component (when $m = 0$). From here comes $N$ in the complexity. The $\log N$ comes from shifting bits or from the DSU operations.

Reply:
that's wrong, operations on a bitset of length N take time proportional to N (but the constant factor is good)

Reply:
Oh, I didn't know that. So it would be $\mathcal{O}(N^2)$ ?